deltacat/compute/compactor_v2/utils/task_options.py

```diff
@@ -0,0 +1,224 @@
+
+    if entry.meta.source_content_length:
+        return entry.meta.source_content_length
+
+    return entry.meta.content_length * previous_inflation
+
+
+def estimate_manifest_entry_num_rows(
+    entry: ManifestEntry,
+    average_record_size_bytes: float,
+    previous_inflation: float,
+    **kwargs,
+) -> int:
+    if (
+        entry.meta
+        and entry.meta.content_type == ContentType.PARQUET
+        and entry.meta.content_encoding == ContentEncoding.IDENTITY
+    ):
+        for type_params in entry.meta.content_type_parameters:
+            if isinstance(type_params, PartialParquetParameters):
+                return type_params.num_rows
+
+    if entry.meta.record_count:
+        return entry.meta.record_count
+
+    total_size_bytes = estimate_manifest_entry_size_bytes(
+        entry=entry, previous_inflation=previous_inflation, **kwargs
+    )
+
+    return int(total_size_bytes / average_record_size_bytes)
+
+
+def estimate_manifest_entry_column_size_bytes(
+    entry: ManifestEntry, columns: Optional[List[str]] = None
+) -> Optional[float]:
+    if not columns:
+        return 0
+
+    if (
+        entry.meta
+        and entry.meta.content_type == ContentType.PARQUET
+        and entry.meta.content_encoding == ContentEncoding.IDENTITY
+    ):
+        for type_params in entry.meta.content_type_parameters:
+            if (
+                isinstance(type_params, PartialParquetParameters)
+                and type_params.pq_metadata
+            ):
+                return _calculate_column_size(type_params=type_params, columns=columns)
+
+    return None
+
+
+def hash_bucket_resource_options_provider(
+    index: int,
+    item: DeltaAnnotated,
+    previous_inflation: float,
+    average_record_size_bytes: float,
+    primary_keys: List[str] = None,
+    **kwargs,
+) -> Dict:
+    size_bytes = 0.0
+    num_rows = 0
+    total_pk_size = 0
+
+    if not item.manifest or not item.manifest.entries:
+        return {"CPU": 0.01}
+
+    for entry in item.manifest.entries:
+        entry_size = estimate_manifest_entry_size_bytes(
+            entry=entry, previous_inflation=previous_inflation
+        )
+        num_rows += estimate_manifest_entry_num_rows(
+            entry=entry,
+            previous_inflation=previous_inflation,
+            average_record_size_bytes=average_record_size_bytes,
+        )
+        size_bytes += entry_size
+
+        if primary_keys:
+            pk_size = estimate_manifest_entry_column_size_bytes(
+                entry=entry,
+                columns=primary_keys,
+            )
+
+            if pk_size is None:
+                total_pk_size += entry_size
+            else:
+                total_pk_size += pk_size
+
+    # total size + pk size + pk hash column + hash bucket index column
+    # Refer to hash_bucket step for more details.
+    total_memory = size_bytes + total_pk_size + num_rows * 20 + num_rows * 4
+
+    # Consider 20% buffer
+    total_memory = total_memory * 1.2
```

Comment:
Could we make the 1.2 buffer as a constant to be more clear and easy to change?

Reply:
done.